Rustdoc LaTeX math (#3958)
rust-lang/rfcs: text/3958-rustdoc-texmath.md

Reviewer: General vibe: I like the approach this takes. I think it takes our stability guarantees into account, and I think we could manage to support it long-term.

I think the *worst-case* failure mode of this is that we decide we have to go back to showing the raw LaTeX syntax rather than rendering it, or that people go back to using external tools like KaTeX; that's not a terrible failure mode.

I also think that if, in the future, we wanted to support some *other* syntax, this wouldn't stop us from doing so. We have the edition mechanism, and we could also support opt-ins for those different possibilities. I added a mention in the alternatives section for `typst`, for instance, along with detailed reasons why we *shouldn't* support `typst` in the near future.

I left a couple of inline comments that arise from the same underlying thought: I think we should defer custom command support, for a few different reasons. (Feel free to respond to that on the thread of those inline comments, rather than here.)

Other than that, this looks good to me, and I'd be happy to see it implemented and shipped. :heart:

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..cb086d6cfa21e88b4cc5cc19790355423b054988)*

- We can report math syntax errors on the CLI, just like we do for intra-doc links.
- We can render math in the resulting web page without JavaScript.
  No flash of unstyled content or blocking scripts.
- Built-in TeX math doesn't require double-escaping, because the Markdown parser knows about math,
  and lets you backslash escape the dollar sign to disable it.
- Cross-crate inlining works.

[internals thread]: https://internals.rust-lang.org/t/adding-latex-support-to-rustdoc/23858
[crates that did it themselves]: https://github.com/search?q=rustdoc-args+%3D+%5B%22--html-in-header%22%2C+%22katex-header.html%22%2C+%22--cfg%22%2C+%22docsrs%22%5D+language%3Atoml&type=code
[katex.js]: https://katex.org/
[include hack]: https://github.com/search?q=%23%21%5Bdoc+%3D+include_str%21%28%22katex.html%22%29%5D+language%3Arust&type=code

## Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

### How to enable

To enable `$\TeX$` math syntax in rustdoc, add this line to your crate root.

    #![doc(math_syntax)]

In a future edition, we may enable it by default. If you need to turn it off, add this line to your crate root.

>     #![doc(no_math_syntax)]

Reviewer 2: This enabling mechanism seems problematic to me with regards to external macros. 

External macro implementers won't be able to use the math syntax, as they won't know if its enabled, since they don't control the crate that enable it.

I don't know what would be a good solution here. Maybe we need a `#[doc]` attribute that one can enable per doc-comment?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..7977d03c61e503eb7e821a8b75d101de4ec7e834)*

Reviewer 3: This seems to me like something that should be settable on any item or module and apply to its contents, like `#[allow]`.

Reviewer 4: Perhaps it could be similar to the unstable `coverage` attribute, like `#[doc(math_syntax = "on")]` and `#[doc(math_syntax = "off")]`, propagated recursively?

That way, adding it to the crate root is simply setting the default for the crate root, and any individual items can enable/disable it.

Author: I would like to support it on arbitrary items, but it would require us do decide how inheritance should work. What, exactly, should this do?

~~~rust
#![doc(math_syntax=(
    first=r”\second”,
    second=“2”,
))]

#[doc(math_syntax(second=r“\text{2nd}”))]
/// $\first$
pub struct Test;
~~~

A crate attribute lets us punt that decision for later. It also lets us see if anyone actually wants to use this, with real use cases. The KaTeX embedding hack that we want to replace doesn’t do this, after all.

Reviewer: @notriddle The desire to do that, and to not deal with the scoping issues you're describing, makes me tempted to defer support for custom functions, for now, in favor of *only* supporting what's built in.

Author: https://github.com/rust-lang/rfcs/pull/3958#discussion_r3175866182

When this feature is enabled, equations are wrapped in single or double `$` dollar signs.

    $$\sum_{i=0}^N x_i$$

The result looks like this:

> $$\sum_{i=0}^N x_i$$

A detailed comparison between our syntax and KaTeX's can be found
[here](https://tmke8.github.io/math-core/comparison.html).

> You can add custom \commands by supplying key=value pairs to the math syntax attribute:

Reviewer: I would like to propose that we defer the custom commands syntax here, and *just* support what ships with stock math-core. This means we don't have to deal with scoping issues, *and* it means we could add more functions ourselves in the future (e.g. upgrade to a version of math-core that implements a larger subset of LaTeX) without worrying about conflict with user-defined functions.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..cb086d6cfa21e88b4cc5cc19790355423b054988)*

Author: I chose this arm of the trade-off triangle (punt on inheritance, have per-item `doc(math_syntax)`, or support custom commands; pick two) because the katex.js hack supports custom commands and when I ran a Crater run with math-core, I found crates[^1] that use that feature.

Since my primary goal here was to get existing crates to switch off the katex,js hack onto this new feature, I prioritized feature parity.

[^1]: In [my report](https://gist.github.com/notriddle/108fe255ffa9f490ed8ade262935eb53#file-report-txt), some "unknown command" errors that are caused by custom commands include: [ket](https://docs.rs/crate/Rustb/latest/source/docs-header.html), [ith](https://docs.rs/crate/cggmp21/latest/source/katex-header.html), [pd, tr](https://docs.rs/crate/fenris/latest/source/assets/doc-before-content.html), [G](https://docs.rs/crate/generic-ec/latest/source/katex-header.html), [sk](https://docs.rs/crate/key-share/latest/source/katex-header.html), [floor](https://docs.rs/crate/ruint/latest/source/.cargo/katex-header.html), and [Zq](https://docs.rs/crate/sn_curv/latest/source/katex-header.html).

Reviewer: @notriddle That is helpful to know. But looking over the list, it seems like most of them are one-offs. I wonder to what extent most crates would be willing to switch even without that feature?

I do think people will want the ability to use this in macros. Which means we're either going to have to punt on custom commands, *or* carefully define inheritance. The latter is *possible* but seems like substantial additional complexity (making it much more a "language" than a predefined set of supported markup).

    #![doc(math_syntax(
        // usage: $\floor{x}$

>         floor=r##"\delim{\lfloor}{#1}{\rfloor}"##,

Reviewer 2: What happens when someone defines a math command but math is disabled? Is it an error?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..7977d03c61e503eb7e821a8b75d101de4ec7e834)*

Author: That's impossible.

The attribute `#![doc(math_syntax)]` is synonymous with `#[doc(math_syntax())]`. This attribute enables math support and adds any commands passed as arguments at the same time.

Reviewer 2: Good, that works for me. Could you make it explicit in the RFC, it was not obvious to me, that this was the case.

>         floor=r##"\delim{\lfloor}{#1}{\rfloor}"##,

Reviewer 5: Could `include!` be made to work within this macro? It could be nice to keep shared definitions in a separate file. (Fine as a future possibility).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..cb086d6cfa21e88b4cc5cc19790355423b054988)*

Reviewer 6: so like `#![doc(math_syntax(include!("my_syntax.txt")))]`? or did you mean `#![doc(math_syntax(foo = include!("foo.txt"), bar = include!("bar.txt")))]`

Reviewer 5: The first one is what I was thinking of, so you could have one file defining custom commands and reuse it for multiple crates in a workspace. (With some symlinking if they're published).

Reviewer 7: this should be possible by making `math_syntax` eagerly expanded, like `doc =` already is.

>     ))]

Reviewer 2: Given how this `math_syntax` sub-attribute is used, I think it should be named something else, something that reflect that one can put "stuff" in it. Maybe simply `#![doc(math)]`?

That would mean:
 - `#![doc(math_syntax)]` -> `#![doc(math)]` (to enable it)
 - `#![doc(no_math_syntax)]` -> `#![doc(no_math)]` (to disable it)
 - ```
   #![doc(math_syntax(
        // usage: $\floor{x}$
        floor=r##"\delim{\lfloor}{#1}{\rfloor}"##,
    ))]
    ```
    ->
    ```
    #![doc(math(
        // usage: $\floor{x}$
        floor=r##"\delim{\lfloor}{#1}{\rfloor}"##,
    ))]
    ```

We could even say that any command definition implicitly enables the math syntax.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..7977d03c61e503eb7e821a8b75d101de4ec7e834)*

Author: Seems to be resolved by https://github.com/rust-lang/rfcs/pull/3958/changes/598e1023d5a7016cd1652e2ebdeb6c05cbf6fee1 ?

## Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

### Disabling and enabling math syntax

The crate-level doc attributes `math_syntax` and `no_math_syntax` enable and disable
support for parsing `$`-delimited TeX math in Rustdoc's Markdown.

Obviously, you can't set both of them at the same time. If neither of them are set,
rustdoc will use the edition-specific default (which is currently to disable it).

The `math_syntax` attribute accepts an optional list of `key="value"` pairs for
custom macros. This is similar to the `macros` parameter that KaTeX accepts,

> but the `key` part only includes the name of the macro, without the backslash

Reviewer 2: ```suggestion
but the `key` is an ident that only includes the name of the macro, without the backslash
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..7977d03c61e503eb7e821a8b75d101de4ec7e834)*

or the number of parameters. The `"value"` is a string literal with TeX-like code,
and, optionally, `#`numbered parameter placeholders.

### Writing math code in markdown

Math expressions are wrapped in `$` signs. One dollar sign means "inline" math,
and two means "display" math.

Inline math cannot have any whitespace at the start or end of its contents,
so `$1$` is a math span, but `$ 1 $` is not. Inline math spans also
can't be empty.

Display math is allowed to have space at the start,
so `$$ 1 $$` is a display math span.

Unescaped curly braces within math spans must balance,
and unescaped dollar signs can only appear between unescaped curly braces,
so `$$ 1 {$} 2 $$` is parsed as a display math span,

> but `$$ 1 $ 2 $$` and `$$ { $$` are not.

Reviewer 8: I don't see much reason why `$$ 1 $ 2 $$` should be invalid, `$$` and `$` are different delimiters. 

Additionally, `$$ { $$` is said to be invalid, but what about like ↓?

```latex
$$ \text{ $$ what? $$ } $$
```

On GitHub this is invalid, the parser only matches a `$$` with another, without considering the balancing of the braces:

> $$ \text{ $$ what? $$ } $$

I'm not sure what this RFC proposes, but IMO it would be easier to also consider this invalid, meaning the parser pairing `$$` to extract math content do not need to first read the content and balance the braces.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..cb086d6cfa21e88b4cc5cc19790355423b054988)*

Author: Our syntax is based on [Pandoc], not GitHub.

[Pandoc]: https://pandoc.org/try/?params=%7B%22text%22%3A%22%24%24+%5C%5Ctext%7B+%24%24+what%3F+%24%24+%7D+%24%24%5Cn%22%2C%22to%22%3A%22json%22%2C%22from%22%3A%22gfm%22%2C%22standalone%22%3Afalse%2C%22embed-resources%22%3Afalse%2C%22table-of-contents%22%3Afalse%2C%22number-sections%22%3Afalse%2C%22citeproc%22%3Afalse%2C%22html-math-method%22%3A%22plain%22%2C%22wrap%22%3A%22auto%22%2C%22highlight-style%22%3Anull%2C%22files%22%3A%7B%7D%2C%22template%22%3Anull%7D

### Math syntax

Within a math span, whitespaces are used for grouping and formatting.
But you can't have more than one line break in a row within a math span,
because that ends the paragraph that contains it.

Other characters are usually rendered literally, except for

- backslashes, `\`, which are the sigil for commands
- curly braces, `{` and `}`, which are used for command arguments
- dollar signs, `$`, which delimit math spans
- number signs, `#`, which are used to refer to macro parameters
- ampersands, `&`, which are used for writing matrices and tables
- circumflex, `^`, which is used for exponents
- underscore, `_`, which is used for subscript
- single quote, `'`, which becomes the prime symbol

> - tile, `~`, which becomes a rendered, non-breaking space (since ordinary spaces are used for grouping)

Reviewer 6: ```suggestion
- tilde, `~`, which becomes a rendered, non-breaking space (since ordinary spaces are used for grouping)
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..48b494e56c2e66d48bd20b5b4391e33515609243)*

- percent, `%`, which mark line comments
- NUL, which is not allowed

Commands are used to write things that can't easily be typed on a keyboard,
and for complex layouts like fractions and matrices. The math-core parser
that we use implements hundreds of commands.

<I>TODO: Full list is in <https://github.com/tmke8/math-core/blob/main/crates/math-core/src/commands.rs>.

> Do I need to include it all here?</I>

Reviewer 2: No, I think we can leave that out. I expect the exact supported syntax to evolve over time.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..7977d03c61e503eb7e821a8b75d101de4ec7e834)*

There are a lot of special-purpose technical notations that we might theoretically want to support,
but TeX-math-in-markdown is special, for two reasons:

- [Lindy effect][]: LaTeX is an established standard that is not going anywhere any time soon.
- There is more than one implementation of the subset of LaTeX that we need.

The pull request I've been working on uses [math-core][], but, if that implementation turns out
to be problematic, we could pivot to another one, like [pulldown-latex][],
or [katex run in quick-js][] [^1].
That’s not an option with, for example, Typst.

[Lindy effect]: https://en.wikipedia.org/wiki/Lindy_effect
[math-core]: https://github.com/tmke8/math-core
[pulldown-latex]: https://github.com/carloskiki/pulldown-latex
[katex run in quick-js]: https://docs.rs/katex/latest/katex/

[^1]:
    I would prefer not to do *that*, because it's slow and seems to have poor error reporting,
    but, if we can't achieve good-enough feature support any other way, it's an option.

> ## Prior art

Reviewer: ```suggestion
### `typst`

Typst is growing in popularity, and is implemented in Rust. It might be an interesting choice to consider in the future. However:
- Typst does not yet have strong support for HTML output; it's still a work in progress. Rustdoc is *primarily* oriented around HTML.
- Typst and Markdown are different in critical ways that may introduce confusion.
- Typst itself is pre-1.0.
- Because typst is newer, it's less immediately obvious which subset we can consider stable to support in the long term, without relying on specific libraries or on execution of typst code in general.

In the future, it might make sense to explore more use of typst in Rust documentation.

## Prior art
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..cb086d6cfa21e88b4cc5cc19790355423b054988)*

[prior-art]: #prior-art

- <https://github.com/cben/mathdown/wiki/math-in-markdown>
- <https://en.wikibooks.org/wiki/LaTeX/Mathematics>
- The span parsing is based on the [math spec for commonmark-hs][],
  which is the parser used if you run `pandoc` in `gfm` mode.
- Span parsing is documented in more detail in the [math spec for pulldown-cmark][].

[math spec for commonmark-hs]: https://github.com/jgm/commonmark-hs/blob/master/commonmark-extensions/test/math.md
[math spec for pulldown-cmark]: https://pulldown-cmark.github.io/pulldown-cmark/specs/math.html

## Unresolved questions
[unresolved-questions]: #unresolved-questions

### Avoiding Hyrum's Law

There are a lot of \commands in [math-core][], and some of them are known buggy,
meaning they don't match LaTeX itself.
We don't want authors to rely on those bugs, either accidentally
or in a workaround.

Normally, we might "phase in" new commands by making them unstable first,
letting more risk-tolerant authors try it out,
then make it available to everyone else.

> But math-core doesn't have an API for that.

Reviewer 5: This makes it sound a bit like some slash commands will be feature gated?

I'm not sure if that's a hoop worth jumping through; I expect that many people writing equations can deal with some level of instability in output for lesser-used features. All math rendering libraries change the way things look as they add features or fix inconsistencies, and browser support for MathML still has places to improve.

Instead, it seems sufficient to say that:

1. We try to make sure the rendered output is reasonably close to what LaTeX does
2. Output that deviates from what LaTeX does may change
3. Output that closely matches LaTeX is unlikely to change

Then if anything, we could keep a list of slash commands that are known to not render completely correctly.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..cb086d6cfa21e88b4cc5cc19790355423b054988)*

Author: I haven't decided whether slash commands should be feature gated or not. That's why it's under the "unanswered questions" section.

### Font

Right now, math formulas default to Noto Sans Math.

This was chosen because it's inoffensive and fine. But it is a sans serif font face, that will usually be surrounded by serif text.

## Future possibilities
[future-possibilities]: #future-possibilities

### Undelimited environments

It's a relatively rare feature, but Jupyter Notebook and a few others support
LaTeX environments introduced with the `\begin{foo}` / `\end{foo}` syntax
without wrapping dollar signs.
Since backslashes in Markdown only have meaning when followed by punctuation,
the false positives shouldn't be that common.
Since we don't have to worry about false positives,
we can treat it like a CommonMark block construct and allow blank lines in it.

    /// Computes sum from `start` to `end` of the given function.
    ///
    /// \begin{equation}
    /// \sum_{i=start}^{end}{f(i)}

>     /// \end{equation}

Reviewer 8: I don't think we should support it. We are primarily writing *Markdown*, not *LaTeX* after all. Bare `\begin{foo}`/`\end{foo}` can just be placed inside `$$ ... $$`.

> ```md
> Computes sum from `start` to `end` of the given function.
>
> $$
> \begin{equation}
> \sum_{i=\mathtt{start}}^{\mathtt{end}}{f(i)}
> \end{equation}
> $$
> ```
>
> Computes sum from `start` to `end` of the given function.
>
> $$
> \begin{equation}
> \sum_{i=\mathtt{start}}^{\mathtt{end}}{f(i)}
> \end{equation}
> $$

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..cb086d6cfa21e88b4cc5cc19790355423b054988)*

Reviewer 4: This isn't super abnormal considering how CommonMark handles block-level HTML, actually. I think it's a neat shortcut to make doc comments slightly shorter, since they are part of the code, after all.

That said, it is not explicitly required here.

- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

## Summary
[summary]: #summary

Add support for the de facto standard TeX-math-in-markdown syntax to Rustdoc. It’s currently implemented using the [math-core][] library, which generates MathML Core and is restricted to the subset of LaTeX that can be implemented that way. If you use unsupported syntax, you get a compiler warning from rustdoc.

[math-core]: https://docs.rs/math-core/latest/math_core/

## Motivation
[motivation]: #motivation

It would be nice if we could write complex equations in our docs.
We know that there's demand for this feature,
first of all because people have [asked for it][internals thread],
but mostly because of [crates that did it themselves][] by loading [katex.js][] with inline HTML.

As far as I know, this is the most popular way of doing that:

    [package.metadata.docs.rs]
    rustdoc-args = ["--html-in-header", "katex-header.html", "--cfg", "docsrs"]

Because only docs.rs reads that directive,
local `cargo doc` and non-Rustdoc doc readers won't see it.
There is a way to make it work in `cargo doc`,
but it [seems to be less popular][include hack].

> Fixing this bug and making it easier to build self-contained docs is

Reviewer 9: > Fixing this bug

Which bug, exactly?

the main motivation for adding built-in support for math syntax to Rustdoc,
but there are a few other quality of life improvements that come with this feature:

- We can report math syntax errors on the CLI, just like we do for intra-doc links.
- We can render math in the resulting web page without JavaScript.
  No flash of unstyled content or blocking scripts.
- Built-in TeX math doesn't require double-escaping, because the Markdown parser knows about math,
  and lets you backslash escape the dollar sign to disable it.
- Cross-crate inlining works.

[internals thread]: https://internals.rust-lang.org/t/adding-latex-support-to-rustdoc/23858
[crates that did it themselves]: https://github.com/search?q=rustdoc-args+%3D+%5B%22--html-in-header%22%2C+%22katex-header.html%22%2C+%22--cfg%22%2C+%22docsrs%22%5D+language%3Atoml&type=code
[katex.js]: https://katex.org/
[include hack]: https://github.com/search?q=%23%21%5Bdoc+%3D+include_str%21%28%22katex.html%22%29%5D+language%3Arust&type=code

## Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

### How to enable

To enable `$\TeX$` math syntax in rustdoc, add this line to your crate root.

>     #![doc(math_syntax)]

Reviewer 9: As Typst is already mentioned below as a future alternative, how would support for it look like w.r.t. this syntax to enable TeX math syntax?

In case Typst proves to be up for the task in future, I suppose having something like `#![doc(math_syntax = "tex")]` now, so that `#![doc(math_syntax = "typst")]` could be added would make sense. However, this would conflict with how custom \commands are defined. Would it make sense to have a different syntax for those, leaving the proposed `#![doc(math_syntax)]` just for declaring the fact that math syntax is in use?

In a future edition, we may enable it by default. If you need to turn it off, add this line to your crate root.

>     #![doc(no_math_syntax)]

Reviewer 9: Generally, what are the use cases for being able to opt out of having math syntax? Backwards compatibility for those already using similar syntax (that's possibly incompatible) without wanting it rendered as TeX?

If a different syntax of using [`#![doc(math_syntax = "tex")]`](https://github.com/rust-lang/rfcs/pull/3958/changes#r3176428391) was used instead for enabling the support, maybe something like `#![doc(math_syntax = "disabled")]` could be used for disabling it.

When this feature is enabled, equations are wrapped in single or double `$` dollar signs.

    $$\sum_{i=0}^N x_i$$

The result looks like this:

> $$\sum_{i=0}^N x_i$$

A detailed comparison between our syntax and KaTeX's can be found
[here](https://tmke8.github.io/math-core/comparison.html).

You can add custom \commands by supplying key=value pairs to the math syntax attribute:

    #![doc(math_syntax(
        // usage: $\floor{x}$
        floor=r##"\delim{\lfloor}{#1}{\rfloor}"##,
    ))]

## Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> ### Disabling and enabling math syntax

Author: @joshtriplett @Urgau 

It looks like neither of you think support for custom \commands is a big deal, you both want to be able to attach the declaration to arbitrary items so that math syntax can be used in macro-emitted doc comments, and you're thinking ahead to replacing markdown entirely with typst syntax.

Here's an alternate design that seems pretty forward-thinking:

> ## Reference-level explanation
>
> ### The `#[doc(syntax="...")]` declaration
>
> To enable math syntax, attach `#[doc(syntax="+tex_math_dollars")]` to an item. To disable it, use `#[doc(syntax="-tex_math_dollars")]`. As a minimum viable initial release, this is the only supported parameter.
>
> The syntax is inherited from the syntactic parent item. So, to list a few examples:
>
> ```rust
> // lib.rs
>
> //! This is *not* math syntax, obviously: $x$
>
> #[doc(syntax="-tex_math_dollars")]
> pub mod bar;
> #[doc(syntax="+tex_math_dollars")]
> pub mod baz;
>
> #[doc(syntax="+tex_math_dollars")]
> /// This *is* math syntax: $x$
> pub struct Foo;
>
> impl Foo {
>     /// This is *not* math syntax: $x$
>     pub fn foo() {}
> }
>
> #[doc(syntax="+tex_math_dollars")]
> impl Foo {
>     /// This *is* math syntax: $x$
>     pub fn bar() {}
> }
> ```
>
> ```rust
> // bar.rs
> #![doc(syntax="+tex_math_dollars")] //~ ERROR only one doc syntax can be declared on a single item
> ```
>
> ```rust
> // baz.rs
> //! This *is* math syntax
> ```
>
> ## Future possibilities
>
> ### Extending `doc(syntax)`
>
> The name `tex_math_dollars` is deliberately the same [name and syntax used by Pandoc](https://pandoc.org/MANUAL.html#extension-tex_math_dollars) for this feature.
>
> If we add more syntactic features, we can follow the same pattern. An example of how to build on this:
>
> ```ebnf
> (* if no language is supplied, the default is "rustdoc_markdown" *)
> doc syntax = [ language ], { ( "+" | "-" ), extension } ;
> (* "commonmark" enables no extensions (you can, of course, add them)
>   "rustdoc_markdown" is synonymous with `commonmark+intra_doc_links+doctests+smart+pipe_tables+strikeout+footnotes+task_lists"
>   "gfm" is synonymous with `commonmark+smart+tex_math_dollars+pipe_tables+strikeout+footnotes+task_lists+emoji+tex_math_gfm+alerts+autolink_bare_uris+yaml_metadata_block"
>   "typst" doesn't support any extensions except "intra_doc_links" and "doctests", so, for example, `doc(syntax="typst+pipe_tables")` is an error *)
> language = "commonmark" | "rustdoc_markdown" | "gfm" | "typst" ;
> extension = 
>    "intra_doc_links"
>  | "doctests"
>  | "smart"
>  | "pipe_tables"
>  | "strikeout"
>  | "footnotes"
>  | "task_lists"
>  | "tex_math_dollars"
>  | "emoji"
>  | "tex_math_gfm"
>  | "alerts"
>  | "autolink_bare_uris"
>  | "yaml_metadata_block" ;
> ```
>
> If a parent and child element both have `doc(syntax)` attributes, they aren't merged. The child just overrides the parent.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..3963d7deb33230198aacfd35743725d5e81f70a1)*

Reviewer 2: That's seems much more forward thinking (both for this feature and new ones we may introduce) and solves the macro-emitted doc comments concern I have. I like it.

This alternate design also gives a clearly a edition-migration strategy. Just put `#[doc(syntax="-tex_math_dollars")]` on every doc-comment containing `$` in the new edition to keep the current behavior.

Reviewer 8: I don't think "*Just put `#[doc(syntax="-tex_math_dollars")]` on every doc-comment containing `$` in the new edition*" is a valid migration strategy in general, because Line 82 specified that "you can't set both of them at the same time on a single item". Meaning a macro usage like

```rust
#[macro_export]
macro_rules! declare {
    ($(#[$meta:meta])* $name:ident) => {
        $(#[$meta])*
        #[doc(syntax="-tex_math_dollars")]
        /// Some extra $non-math$ docs.
        pub struct $name;
    }
}

// ... in another crate ...

external_crate::declare! {
    #[doc(syntax="+tex_math_dollars")]
    /// Hello $3^2+4^2=5^2$
    Foo
}
```

would be a rustdoc-compilation error.

Reviewer 10: Linking to @<span></span>kennytm's [follow-up](https://github.com/rust-lang/rfcs/pull/3958#discussion_r3252279701) (could be sent right here :))

Author: Put `#![doc(syntax="-tex_math_dollars")]` on the crate, unless this attribute is already set there, if any of the doc comments have dollar signs in them.

Reviewer 10: Does this mean that crate-level `#[doc(syntax)]` dominates? (Is it still possible to have math on crate items unaffected by problematic macro use?)

Author: It’s inherited. If a child item has doc(syntax), it’ll override the parent. So, setting it on the crate will establish a default.

[^ghmath]:
    Did GitHub run a similar batch job when they added math syntax?
    I can't think of any reason why they wouldn't, but I also can't find any proof that they did.
    It seems like it would require running the math-enabled parser over all the issue comments,
    and, if it detects math, add a backslash in front of the dollar signs.
    After all, math syntax didn't exist in GitHub Issues until they added it,
    so any detected math span is, by definition, a false positive.

[CommonMark spec]: https://spec.commonmark.org/0.31.2/#characters-and-lines
[one-time batch converter job]: https://github.blog/engineering/a-formal-spec-for-github-markdown/#the-migration
[intra-doc links were designed]: https://github.com/rust-lang/rust/issues/54191
[pulldown-cmark was last updated]: https://github.com/rust-lang/rust/pull/121659#issuecomment-1992752820
[hoedown was replaced with pulldown-cmark in the first place]: https://internals.rust-lang.org/t/what-to-do-about-pulldown-and-commonmark/5115
[anyone proposes replacing Markdown with something else]: https://internals.rust-lang.org/t/rustdoc-restructuredtext-vs-markdown/356

### Verbosity or breakage as side effect

> From the perspective of 99% of doc authors who didn't want to write a math span in the first place,

Reviewer 9: > From the perspective of 99% of doc authors

Is this an actual statistic or a conjecture?

Author: It's conjecture. There aren't very many crates that use the katex hack, but that could be caused by the a11y problems and poor devex. What we care about is the number of crates that *would* use this feature if it were built in, and we can't know that.

I suspect that few crates would use this feature even if it were built in, because most crater runs I've run turn up a bunch of crates where the author wrote plain text without accounting for Markdown's quirks. When I ran a [crater run](https://gist.github.com/notriddle/108fe255ffa9f490ed8ade262935eb53#file-report-txt) to find problems caused by adding LaTeX math support, I found a lot of crates that had `$` signs in code that wasn't escaped. When I did the [crater run to update pulldown-cmark](https://github.com/rust-lang/rust/pull/121659#issuecomment-1992752820), I found a bunch of "Questionable" code that was broken with both versions, because they had unescaped ASCII art.

As a gut check, I just ran `cargo lintcheck` on the top 200 crates with [a custom lint](https://github.com/notriddle/rust-clippy/commit/ae12ea48e) to measure feature usage. *This isn't measuring the same thing as crater,* but it's still interesting:

| number of crates | feature |
|-|-|
| [10](https://gist.github.com/notriddle/07890ef4cd8fa6264c1472ef6550f619#file-has_footnote-txt) (5%) | footnotes
| [21](https://gist.github.com/notriddle/07890ef4cd8fa6264c1472ef6550f619#file-has_table-txt) (10%) | tables
| [46](https://gist.github.com/notriddle/07890ef4cd8fa6264c1472ef6550f619#file-has_html-txt) (23%) | HTML
| [143](https://gist.github.com/notriddle/07890ef4cd8fa6264c1472ef6550f619#file-has_list-txt) (71%) | lists (bullet or numbered)
| [176](https://gist.github.com/notriddle/07890ef4cd8fa6264c1472ef6550f619#file-has_link-txt) (88%) | links

Reviewer 9: Thanks for clarifying.

> What we care about is the number of crates that would use this feature if it were built in, and we can't know that.

FWIW, at my workplace, we haven't bothered making math syntax render nicely specifically because of the feature not existing in vanilla rustdoc. We'd fix that as soon as the feature became available, and any backwards incompatibilities due to the dollar signs would just be shrugged off as well worth it.

so Rustdoc compiler warnings won't catch every accidental match.

But if we assume that every doc author adds the escapes that they need,
this forces doc comments to have more escaped metacharacters than they used to.
This makes doc comments less easily readable in their source form,
imposing a cost on the 99% that don't want the feature in favor of the 1% who do.

This argument, if taken to its logical extreme, would imply that we should use plain text
doc comments with no extra formatting features. The downside of doing that is
similar to the downside of not offering TeX math: users who *really* want bold text deploy
[unicode crimes][] and pictures of text, which create accessibility problems.

[accidentally triggering a link refdef]: https://github.com/rust-lang/rust/issues/133150
[example rendering of a mistake]: https://tmke8.github.io/math-core/#input:H4sIAAAAAAAAEwXBwQ0AIQgEwFa2guvCQvaBSoJIhETLv5n2uMIEu6P5MM2JODsF92iVONRBGEseFmsig_79a-7ZDjYAAAA=
[unicode crimes]: https://ux.stackexchange.com/questions/118149/can-screen-readers-interpret-unicode-styles-fonts-such-as-bold-and-italics

## Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> 

Reviewer 9: It would be prudent to give a rationale for the specific approaches of:

- Rendering in the "backend" rather than in browser (good reasons were already listed under Motivation)
- Using MathML which seems to be quite well supported: https://caniuse.com/mathml

  which is the parser used if you run `pandoc` in `gfm` mode.
- Span parsing is documented in more detail in the [math spec for pulldown-cmark][].

[math spec for commonmark-hs]: https://github.com/jgm/commonmark-hs/blob/master/commonmark-extensions/test/math.md
[math spec for pulldown-cmark]: https://pulldown-cmark.github.io/pulldown-cmark/specs/math.html

## Unresolved questions
[unresolved-questions]: #unresolved-questions

### Avoiding Hyrum's Law

There are a lot of \commands in [math-core][], and some of them are known buggy,
meaning they don't match LaTeX itself.
We don't want authors to rely on those bugs, either accidentally
or in a workaround.

Normally, we might "phase in" new commands by making them unstable first,
letting more risk-tolerant authors try it out,
then make it available to everyone else.
But math-core doesn't have an API for that.

### Font

Right now, math formulas default to Noto Sans Math.

This was chosen because it's inoffensive and fine. But it is a sans serif font face, that will usually be surrounded by serif text.

## Future possibilities
[future-possibilities]: #future-possibilities

### Extending `doc(syntax)`

> The name `tex_math_dollars` is deliberately the same [name and syntax used by Pandoc](https://pandoc.org/MANUAL.ml#extension-tex_math_dollars) for this feature.

Reviewer 8: ```suggestion
The name `tex_math_dollars` is deliberately the same [name and syntax used by Pandoc](https://pandoc.org/MANUAL.html#extension-tex_math_dollars) for this feature.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..57a284afc632d44aecc50848472156213ac2586a)*

- Feature Name: `rustdoc_texmath`
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#3958](https://github.com/rust-lang/rfcs/pull/3958)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

## Summary
[summary]: #summary

Add support for the de facto standard TeX-math-in-markdown syntax to Rustdoc. It’s currently implemented using the [math-core][] library, which generates MathML Core and is restricted to the subset of LaTeX that can be implemented that way. If you use unsupported syntax, you get a compiler warning from rustdoc.

[math-core]: https://docs.rs/math-core/latest/math_core/

## Motivation
[motivation]: #motivation

It would be nice if we could write complex equations in our docs.
We know that there's demand for this feature,
first of all because people have [asked for it][internals thread],
but mostly because of [crates that did it themselves][] by loading [katex.js][] with inline HTML.

As far as I know, this is the most popular way of doing that:

    [package.metadata.docs.rs]
    rustdoc-args = ["--html-in-header", "katex-header.html", "--cfg", "docsrs"]

Because only docs.rs reads that directive,

> local `cargo doc` and non-Rustdoc doc readers won't see it.

Reviewer 11: it could work for `cargo`. see this:

https://github.com/victe/rust-latex-doc-minimal-example/blob/1f13abc8d6e84a61e6f5bfe130981a1b83d552be/readme.md?plain=1#L50

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..b5db067728e30c4ded9d8f9abe246b440602eee6)*

Author: This crate has a [`.cargo/config.toml`](https://docs.rs/crate/rust-latex-doc-minimal-example/latest/source/.cargo/config.toml) file with those options in it.

But, if I depend on that crate as a library, then run `cargo doc` from my project, `rust-latex-doc-minimal-example`'s math isn't rendered:

<img width="2880" height="1576" alt="image" src="https://github.com/user-attachments/assets/c516bb32-c51a-4811-818c-69d0852cce8b" />

Reviewer 11: does latex in your own docs render correctly?

Author: My docs don’t use math. Only my dependency does.

Author: I’ve added a note about this to the RFC.

[guide-level-explanation]: #guide-level-explanation

### How to enable

To enable `$\TeX$` math syntax in rustdoc, add this line to your crate root.

    #![doc(syntax="+tex_math_dollars")]

In a future edition, we may enable it by default. If you need to turn it off, add this line to your crate root.

    #![doc(syntax="-tex_math_dollars")]

The only supported values for `doc(syntax)` are `"+tex_math_dollars"` and `"-tex_math_dollars"`,
but we left the option open for future extensions.

When this feature is enabled, equations are wrapped in single or double `$` dollar signs.

    $$\sum_{i=0}^N x_i$$

The result looks like this:

> $$\sum_{i=0}^N x_i$$

A detailed comparison between our syntax and KaTeX's can be found
[here](https://tmke8.github.io/math-core/comparison.html).

## Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

### Disabling and enabling math syntax

The `doc(syntax="+tex_math_dollars"|"-tex_math_dollars")` attribute enables and disables
support for parsing `$`-delimited TeX math in Rustdoc's Markdown.

Obviously, you can't set both of them at the same time on a single item.
If neither of them are set,

> rustdoc will use the edition-specific default (which is currently to disable it).

Reviewer 8: Further follow up on https://github.com/rust-lang/rfcs/pull/3958#discussion_r3252254961. Allowing each edition able to have different default `doc(syntax)` means that the documentation parser have to be able to switch syntax between spans for a single item, e.g. to allow 

```rust
macro_rules! crate_2027_inner {
    ($item:item) => {
        /// Extra doc with $*\LaTeX*$ support
        $item
    }
}

// ...

macro_rules! crate_2024_outer {
    ($item:item) => {
        /// Extra doc with $*no-math*$
        crate_2027::crate_2027_inner! { 
            $item
        }
    }
}

// ...

crate_2024::crate_2024_outer! {
    pub struct Foo;
}
```

to generate an AST in the form like

```xml
<section>
  <p>Extra doc with $<em>no-math</em>$</p>
  <p>Extra doc with <tex-math mode="inline">*\LaTeX*</tex-math> support</p>
</section>
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..13e3e1d65285f750046643f11395eac34cf9018b)*

pub mod baz;

#[doc(syntax="+tex_math_dollars")]
/// This *is* math syntax: $x$
pub struct Foo;

impl Foo {
    /// This is *not* math syntax: $x$
    pub fn foo() {}
}

#[doc(syntax="+tex_math_dollars")]
impl Foo {
    /// This *is* math syntax: $x$
    pub fn bar() {}
}

#[doc(syntax="+tex_math_dollars")]
/// <https://github.com/rust-lang/rust/blob/29155a4cd6/src/librustdoc/visit_ast.rs#L454>
pub fn wizzywig() {
    #[doc(syntax="-tex_math_dollars")] //~ WARN doc syntax can only be declared on items that actually appear in documentation
    pub struct WizzyWig;

    impl Foo {
        /// This *is* math syntax: $x$
        pub fn baz() {}
    }
}
```

```rust
// bar.rs
#![doc(syntax="+tex_math_dollars")] //~ ERROR only one doc syntax can be declared on a single item
```

```rust
// baz.rs
//! This *is* math syntax
```

### Writing math code in markdown

Math expressions are wrapped in `$` signs. One dollar sign means "inline" math,
and two means "display" math.

Inline math cannot have any whitespace at the start or end of its contents,
so `$1$` is a math span, but `$ 1 $` is not. Inline math spans also
can't be empty.

Display math is allowed to have space at the start,

> so `$$ 1 $$` is a display math span.

Reviewer 8: I think we should clarify that the starting `$`/`$$` must not be preceded by a word character, and similarly the ending `$`/`$$` must not be followed by a word character.

This avoids `$1 and US$2` being formatted as tex_math_dollars.

Note that this is stricter than pandoc's behavior, which will math-ify `CA$x.xx and US$y.yy`, but I think it's not a desirable behavior. 

GitHub also won't turn this into an equation.

> ```md
> CA$x.xx and US$y.yy
> ```
>
> CA$x.xx and US$y.yy

You wrap a `<span>` if you actually need to treat that part as TeX

> ```md
> CA<span>$x.xx and US$</span>y.yy
> ```
> 
> CA<span>$x.xx and US$</span>y.yy

Reviewer 12: Personally, I would find this behavior surprising, since there are legitimate uses for putting math right next to text, such as `$n$th` → <span>$n$</span>th. But if this is the default GFM behavior, I guess that deviating from it would also be surprising to some people.

Author: I had a similar argument on the commonmark-hs tracker. I argued for kennytm's side, and jgm argued for bluebear94's side. I didn't win.

If possible, I'd prefer to keep our behavior the same as commonmark-hs's behavior, since that's the parser pandoc uses, and processing rustdoc's markdown using pandoc is very useful. Also, jgm knows more about markdown than I do, so I'll defer to his experience.

Reviewer 8: @bluebear94 the pandoc behavior on delimiters this RFC proposed already deviated from GFM actually https://github.com/rust-lang/rfcs/pull/3958#discussion_r3174354855. But since rustdoc (`pulldown-cmark`) is neither pandoc or GFM we have the freedom to copy and mix whichever behavior.

Also it could be spelled $n$-th.

Author: We decided on (commonmark-hs and pulldown-cmark both merged) a rule where the closing `$` can’t be followed by a decimal digit, but can be followed by other characters. So `US$1.00 or CA$1.30` won’t parse as a math expression, but `$\pi$th` will. This is the same rule comrak uses.

[one-time batch converter job]: https://github.blog/engineering/a-formal-spec-for-github-markdown/#the-migration
[intra-doc links were designed]: https://github.com/rust-lang/rust/issues/54191
[pulldown-cmark was last updated]: https://github.com/rust-lang/rust/pull/121659#issuecomment-1992752820
[hoedown was replaced with pulldown-cmark in the first place]: https://internals.rust-lang.org/t/what-to-do-about-pulldown-and-commonmark/5115
[anyone proposes replacing Markdown with something else]: https://internals.rust-lang.org/t/rustdoc-restructuredtext-vs-markdown/356

### Verbosity or breakage as side effect

From the perspective of 99% of doc authors who didn't want to write a math span in the first place,
false positives that mangle their generated docs are a nasty papercut.
Failing to escape the dollar signs when you needed to is not as bad as [accidentally triggering a link refdef][],
since the degraded result might still be [legible][example rendering of a mistake],
but having to read English text without any spaces sucks.
Also, the LaTeX math syntax is forgiving enough that normal text is often valid,

> so Rustdoc compiler warnings won't catch every accidental match.

Reviewer 10: ```suggestion
Having to read English text without any spaces caused by failing to escape the dollar signs sucks
(though not as bad as [accidentally triggering a link refdef][],
as [the result][example rendering of a mistake] might still be legible).
The LaTeX math syntax is forgiving enough that English prose is often syntactically valid LaTeX,
so Rustdoc invalid LaTeX warnings can't catch every unescaped dollar sign.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..13e3e1d65285f750046643f11395eac34cf9018b)*

Author: Yeah, your version really does flow better.

But if we assume that every doc author adds the escapes that they need,
this forces doc comments to have more escaped metacharacters than they used to.
This makes doc comments less easily readable in their source form,

> imposing a cost on the 99% that don't want the feature in favor of the 1% who do.

Reviewer 10: [todo] This argument sounds invalid to me []

Is this referring to what happens after/when the default is to enable math syntax? It wouldn't make sense otherwise, since if math syntax

This argument, if taken to its logical extreme, would imply that we should use plain text
doc comments with no extra formatting features. The downside of doing that is
similar to the downside of not offering TeX math: users who *really* want bold text deploy
[unicode crimes][] and pictures of text, which create accessibility problems.

[accidentally triggering a link refdef]: https://github.com/rust-lang/rust/issues/133150
[example rendering of a mistake]: https://tmke8.github.io/math-core/#input:H4sIAAAAAAAAEwXBwQ0AIQgEwFa2guvCQvaBSoJIhETLv5n2uMIEu6P5MM2JODsF92iVONRBGEseFmsig_79a-7ZDjYAAAA=
[unicode crimes]: https://ux.stackexchange.com/questions/118149/can-screen-readers-interpret-unicode-styles-fonts-such-as-bold-and-italics

## Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> ### Why TeX math in markdown, specifically?

Reviewer 13: I fully agree with the choice of LaTeX, but for completeness it may be worthwhile to mention a few other alternatives:

## [AsciiMath](https://asciimath.org/index-mathml.html)

### Pros

- Small and simple
- Many existing implementations, including several in Rust
- Easy to write, relatively easy to read, less verbose than TeX

### Cons

- Fewer features than TeX
- Not as widespread
- No standard syntax for embedding in Markdown
- Similar to Markdown, because there are no invalid documents, updates tend to be breaking

## [UnicodeMath](https://www.unicode.org/notes/tn28/)

### Pros

- Uses Unicode, so the plain-text representation more closely resembles the rendered output
- Simple cases are very readable
- Has a standard syntax for Markdown embedding (between `⁅` … `⁆` brackets) which is unlikely to break existing docs

### Cons

- Hard to type, cannot use standard computer keyboard
- Complicated, large spec
- Very complex formulas are often less readable than alternatives
- Not many implementations, and none in Rust
- Not widely used, major user is Microsoft Office

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/17a311afa36c353e88578e9dde44225aaa447ee1..b5db067728e30c4ded9d8f9abe246b440602eee6)*

Author: I added a full comparison table to this section.

### Drawing and charting syntax

There are a lot of different chart formats we *could* try to support.
The tough part is that we want to support it long-term,
give error messages at compile time (if the language has a concept of errors),
and, ideally, have a specification without much churn.

- [PlantUML][] is pretty much exactly what we would want.
  But we don't want to bundle a JRE.
- The other obvious choice is [Mermaid][], because GitHub supports it.
  The upside is that it's popular and terse. The downside is that the only existing
  implementation is a JavaScript library. We could copy in the JS library and embed
  the source code into our HTML, but we wouldn't be able to give syntax errors at
  Rustdoc compile time that way.
- If we supported undelimited LaTeX environment blocks, then it would make sense to
  implement a subset of the [LaTeX drawing tools][] on top of SVG.
  The downside is that the only other implementation of these languages that I know of is [LaTeXML][], which is not written in Rust.
  The upside is that integrating with the math engine means you can embed equations in your charts,

>   and if ee add support for TeX macros, those should work, too.

Reviewer 13: ```suggestion
  and if we add support for TeX macros, those should work, too.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/36463eefe8d1cd0ada0a19350939a22d3d731656..008970f2501e5b0ef7e6915a5e00ebd8399a500d)*

But if we assume that every doc author adds the escapes that they need,
this forces doc comments to have more escaped metacharacters than they used to.
This makes doc comments less easily readable in their source form,
imposing a cost on the 99% that don't want the feature in favor of the 1% who do.

This argument, if taken to its logical extreme, would imply that we should use plain text
doc comments with no extra formatting features. The downside of doing that is
similar to the downside of not offering TeX math: users who *really* want bold text deploy
[unicode crimes][] and pictures of text, which create accessibility problems.

[accidentally triggering a link refdef]: https://github.com/rust-lang/rust/issues/133150
[example rendering of a mistake]: https://tmke8.github.io/math-core/#input:H4sIAAAAAAAAEwXBwQ0AIQgEwFa2guvCQvaBSoJIhETLv5n2uMIEu6P5MM2JODsF92iVONRBGEseFmsig_79a-7ZDjYAAAA=
[unicode crimes]: https://ux.stackexchange.com/questions/118149/can-screen-readers-interpret-unicode-styles-fonts-such-as-bold-and-italics

## Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> ### Why LaTeX math in markdown, specifically?

Author: Moving https://github.com/rust-lang/rfcs/pull/3958#issuecomment-4792291641 to a thread:

> I'm happy to see LaTeX support in Rustdoc moving forward! I wanted to chime in as a maintainer as KaTeX, but an outsider (so far) to the Rust community, so feel free to ignore these comments if they don't match the standards of Rust and/or Rustdoc. I might also be biased here as a maintainer of one of the "competitors" to math-core, but my goal is to be neutral.

Thank you for leaving your feedback! I've been hoping for better, more educated guesses about how this feature would be used if it were rolled out.

I want to start out with a minimal viable product:

- It needs to be minimal, because a lot of this design is going to be based on guesswork.
- It needs to be viable, because if nobody uses it then we won't get any feedback to improve our guesses.

This means leaving out features that we can add later, but not stripping the features back so much that it's unusable. That's a tough balance, and we need all the help we can get.

> I'd like the RFC to leave room for (either now or in the future) for alternate LaTeX rendering backends.

Sure, the current writing in the RFC doesn't make it impossible for us to add a `#[doc(math_engine="katex")]` option later. The RFC prescribes math-core right now, but the "Why LaTeX math in markdown, specifically?" section spells out that we should be able to pivot to a new engine in the future, and that pivot can include adding a way to change it.

If other @rust-lang/rustdoc members favor this, then I can add this as a "Future possibility". But, I, personally, don't think it's a good idea.

> I agree that math-core is probably the right default, being written in Rust, but as the RFC says, "math-core [...] generates MathML Core and is restricted to the subset of LaTeX that can be implemented that way". I'd personally rather use KaTeX (having spent a lot of time making it as compatible as possible with LaTeX), and others might prefer Temml or MathJax (though that's harder to use server-side, it's still possible in MathML or SVG modes).

To understand the problem with this request, I think it's helpful to split "users" down into three distinct [user roles](https://gist.github.com/notriddle/c289e77f3ed469d1c0238d1d135d49e1):

- An "author" writes doc comments and Rust code in a crate
- A "reader" consumes those comments in some way
- A "publisher" is an optional middleman that uses the rustdoc CLI to produce HTML that they upload to a server somewhere

I think the choice of math renderer should be left up to publishers and readers, because different math renderers support different output formats. Rust-Analyzer, in particular, is an officially supported tool, and is at the mercy of whatever math renderer the host IDE has. Other tools, like [rustdoc-md](https://docs.rs/crate/rustdoc-md/latest) and [term-rustdoc](https://github.com/zjp-cn/term-rustdoc), can't run arbitrary JS math renderers either, and even though they aren't supported right now, I don't want to condemn them to never be supported.

But, since you're talking about features that katex has and math-core doesn't, I think you're asking for a way for *authors* to choose their math renderer.

> I can see the advantage to having a single standard such as math-core, because then everyone can code to that standard, and if something important is missing, you could try to add it to math-core and everyone benefits.

My goal, in the design of this feature, is to fix the reader-side problems with the docs.rs-specific JS hacks. If you stick with supported features, fix any warnings that rustdoc or clippy give you regarding your doc comments, and don't do anything "obviously" bad (like using `<script>` tags), your docs should work in users' IDEs.

That means authors need to write their doc comments in a lowest-common-denominator of the popular math renderers.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/36463eefe8d1cd0ada0a19350939a22d3d731656..46bbace54d2dae44b45a79404f42b4eb9d05928a)*

the main motivation for adding built-in support for math syntax to Rustdoc,
but there are a few other quality of life improvements that come with this feature:

- We can report math syntax errors on the CLI, just like we do for intra-doc links.
- We can render math in the resulting web page without JavaScript.
  No flash of unstyled content or blocking scripts.
- Built-in TeX math doesn't require double-escaping, because the Markdown parser knows about math,
  and lets you backslash escape the dollar sign to disable it.
- Cross-crate inlining works.

[internals thread]: https://internals.rust-lang.org/t/adding-latex-support-to-rustdoc/23858
[crates that did it themselves]: https://github.com/search?q=rustdoc-args+%3D+%5B%22--html-in-header%22%2C+%22katex-header.html%22%2C+%22--cfg%22%2C+%22docsrs%22%5D+language%3Atoml&type=code
[katex.js]: https://katex.org/
[include hack]: https://github.com/search?q=%23%21%5Bdoc+%3D+include_str%21%28%22katex.html%22%29%5D+language%3Arust&type=code

## Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

### How to enable

To enable `$\TeX$` math syntax in rustdoc, add this line to your crate root.

>     #![doc(syntax="+tex_math_dollars")]

Reviewer 14: As far as I know, this syntax is unique. Why not going for the "usual" `enable()`/`disable()`?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/f635361c8e3d927208afd6097d4fc5f30e703b43..386e8ed568b76f29312c5c94d63fb8f4f1ed0cf8)*

Reviewer 2: It's not that unique, even in the Rust world, it's used by `-C target-feature=+avx2,-sse4.1`, but it sure is new and unique for an attribute.

`#[doc(syntax(enable(tex_math_dollars)))]` and `#[doc(syntax(disable(tex_math_dollars)))]` seems fine to me.

Author: That sure is a lot of parenthesis, but sure.

Reviewer 15: Yeah, it's a lot of parentheses, but I agree it's better to use a more standard syntax rather than have to deal with parsing pluses and minuses.

[^1]:
    I would prefer not to do *that*, because it's slow and seems to have poor error reporting,
    but, if we can't achieve good-enough feature support any other way, it's an option.

<table>
<caption>Comparison with other math equation formats</caption>
<thead>
<tr>
  <th>Name</th>
  <th>Release date</th>
  <th>Rust implementation</th>
  <th>Implementation-agnostic specification</th>
  <th>Markdown embedding standard</th>
  <th>Easy to type on a keyboard</th>
  <th>Popularity</th>
</tr>
</thead>
<tbody>
<tr>
  <th>LaTeX</th>
  <td>1984</td>
  <td>✅</td>
  <td>❎</td>
  <td>✅</td>
  <td>✅</td>
  <td>High</td>
</tr><tr>
  <th>Typst</th>
  <td>2019</td>
  <td>✅</td>
  <td>❎</td>
  <td>❎</td>
  <td>✅</td>
  <td>Moderate</td>
</tr><tr>
  <th>AsciiMath</th>
  <td>2009</td>
  <td><a href="https://docs.rs/asciimath-rs/latest/asciimath_rs/">✅</a></td>
  <td><a href="https://asciimath.org/index-mathml.html#syntax">✅</a></td>
  <td>❎</td>
  <td>✅</td>
  <td>Moderate</td>
</tr><tr>
  <th>UnicodeMath</th>
  <td>2006</td>
  <td><a href="https://murrayiii.github.io/UnicodeMathML/playground/">❎</a></td>
  <td><a href="https://www.unicode.org/notes/tn28/">✅</a></td>
  <td>✅</td>
  <td>❎</td>

>   <td>I would need access to <a href="https://support.microsoft.com/en-us/office/linear-format-equations-using-unicodemath-and-latex-in-word-2e00618d-b1fd-49d8-8cb4-8d17f25754f8">Microsoft Office</a> feature analytics to answer this question.</td>

Reviewer 5: This is the smallest of nits but ✅ and ❎ are kind of hard to tell apart at a glance. At first I thought the table was saying that all four languages pass all four criteria 🙂 

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/f635361c8e3d927208afd6097d4fc5f30e703b43..7acc3999b43598b01395bb8a67055cf01daf7ae9)*

Author: Is ❌ more readable?

Reviewer 5: Very much so :)

    fn sum(f: impl FnMut(usize) -> usize, start: usize, end: usize) -> usize {
        let mut result = 0;
        for i in start..=end {
            result += f(i);
        }
        result
    }

### Custom macros

Some implementations, such as [Pandoc's `latex_macros` extension][],
let you write LaTeX macro definitions directly in Markdown.
For example, these two doc comments are equivalent:

```rust
#[doc(syntax(enable(tex_math_dollars, latex_macros)))]
/// \newcommand{\tuple}[1]{\langle #1 \rangle}
///
/// $\tuple{a, b, c}$
fn foo_bar() {}

#[doc(syntax(enable(tex_math_dollars, latex_macros)))]
/// $\langle a, b, c \rangle$
fn foo_bar() {}
```

Existing users of rustdoc and katex already write custom functions.
This isn't a niche feature.

[Pandoc's `latex_macros` extension]: https://pandoc.org/MANUAL.html#extension-latex_macros

### Drawing and charting syntax

There are a lot of different chart formats we *could* try to support.
The tough part is that we want to support it long-term,
give error messages at compile time (if the language has a concept of errors),
and, ideally, have a specification without much churn.

- [PlantUML][] is pretty much exactly what we would want.

>   But we don't want to bundle a JRE.

Reviewer 5: _Technically_ it's also js https://github.com/plantuml/plantuml#-building-for-the-browser-teavm, though of course that doesn't change much.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3958/f635361c8e3d927208afd6097d4fc5f30e703b43..7acc3999b43598b01395bb8a67055cf01daf7ae9)*
